### ISSUE TITLE
Stack upgrade 2.1.3 -> 2.3.1 failing with `stack executable not found`

### ISSUE THREAD

**Issue by AlistairB** [commercialhaskell/stack#5283]
### General summary/comments

Stack upgrade is failing on my host PC, Ubuntu 20.04 going from 2.1.3 to 2.3.1 with:

```bash
$ stack upgrade --binary-only
Current Stack version: 2.1.3, available download version: 2.3.1
Newer version detected, downloading
Querying for archive location for platform: linux-x86_64-static
Downloading from: https://github.com/commercialhaskell/stack/releases/download/v2.3.1/stack-2.3.1-linux-x86_64-static.tar.gz
Stack executable "stack-2.3.1-linux-x86_64-static/stack" not found in archive from https://github.com/commercialhaskell/stack/releases/download/v2.3.1/stack-2.3.1-linux-x86_64-static.tar.gz
CallStack (from HasCallStack):
  error, called at src/Stack/Setup.hs:2012:33 in stack-2.1.3-HEoMsL9gvgt2vNexTWWwGx:Stack.Setup
```

I get the same issue when upgrading via docker on debian stretch / buster.

### Steps to reproduce

For example:

1. With docker, run `docker run haskell:8.8.3@sha256:7c8ff5c85fb0ad9ab1c63f578c31a52e5b9ce01d62e669702fba8dee2fa5cd30 stack upgrade --verbose --binary-only --binary-version 2.3.1`

This docker image is debian stretch based with stack 2.1.3

### Expected

Stack to upgrade successfully to 2.3.1

### Actual

```
$ docker run haskell:8.8.3@sha256:7c8ff5c85fb0ad9ab1c63f578c31a52e5b9ce01d62e669702fba8dee2fa5cd30 stack upgrade --verbose --binary-only --binary-version 2.3.1
docker run haskell:8.8.3@sha256:7c8ff5c85fb0ad9ab1c63f578c31a52e5b9ce01d62e669702fba8dee2fa5cd30 stack upgrade --verbose --binary-only --binary-version 2.3.1
Version 2.1.3, Git revision 0fa51b9925decd937e4a993ad90cb686f88fa282 (7739 commits) x86_64 hpack-0.31.2
2020-04-29 22:55:21.979302: [debug] No project config file found, using defaults.
2020-04-29 22:55:21.985279: [debug] SELECT COUNT(*) FROM "last_performed" WHERE ("action"=?) AND ("timestamp">=?); [PersistInt64 1,PersistUTCTime 2020-04-28 22:55:21.985257 UTC]
2020-04-29 22:55:21.985885: [debug] SELECT "id","action","timestamp" FROM "last_performed" WHERE "action"=?; [PersistInt64 1]
2020-04-29 22:55:21.986035: [debug] UPDATE "last_performed" SET "timestamp"=? WHERE "id"=? ; [PersistUTCTime 2020-04-29 22:55:21.985257 UTC,PersistInt64 1]
2020-04-29 22:55:21.986207: [debug] SELECT "id", "action", "timestamp" FROM "last_performed" WHERE "id"=? ; [PersistInt64 1]
2020-04-29 22:55:22.469872: [info] Current Stack version: 2.1.3, available download version: 2.3.1
2020-04-29 22:55:22.470234: [info] Newer version detected, downloading
2020-04-29 22:55:22.470342: [info] Querying for archive location for platform: linux-x86_64-static
2020-04-29 22:55:22.470452: [info] Downloading from: https://github.com/commercialhaskell/stack/releases/download/v2.3.1/stack-2.3.1-linux-x86_64-static.tar.gz
2020-04-29 22:55:26.826586: [error] Stack executable "stack-2.3.1-linux-x86_64-static/stack" not found in archive from https://github.com/commercialhaskell/stack/releases/download/v2.3.1/stack-2.3.1-linux-x86_64-static.tar.gz
CallStack (from HasCallStack):
  error, called at src/Stack/Setup.hs:2012:33 in stack-2.1.3-4VOPUF2kWRK7hcgX69ZWGe:Stack.Setup
```

### Stack version

```
$ stack --version
Version 2.1.3, Git revision 0fa51b9925decd937e4a993ad90cb686f88fa282 (7739 commits) x86_64 hpack-0.31.2
```

### Method of installation

`curl -sSL https://get.haskellstack.org/ | sh`

**Comment by tommilligan** [comment]
I also see this error when upgrading during a docker build based on `ubuntu:18.04`.

The issue appears to be the tarball contains a directory with the name `stack-2.3.1-linux-x86_64`, but it is expected to be `stack-2.3.1-linux-x86_64-static`.

This seems to be likely a result of the reduced platform support and/or change of stable binary locations, as mentioned in the [most recent changelog](https://github.com/commercialhaskell/stack/blob/master/ChangeLog.md#v231)?

**Comment by AlistairB** [comment]
This looks to be fixed (for debian at least.

```bash
$ docker run haskell:8.8.3@sha256:7c8ff5c85fb0ad9ab1c63f578c31a52e5b9ce01d62e669702fba8dee2fa5cd30 stack upgrade --verbose --binary-only --binary-version 2.3.1

Version 2.1.3, Git revision 0fa51b9925decd937e4a993ad90cb686f88fa282 (7739 commits) x86_64 hpack-0.31.2
2020-05-04 04:20:28.567201: [debug] No project config file found, using defaults.
2020-05-04 04:20:28.577697: [debug] SELECT COUNT(*) FROM "last_performed" WHERE ("action"=?) AND ("timestamp">=?); [PersistInt64 1,PersistUTCTime 2020-05-03 04:20:28.577672 UTC]
2020-05-04 04:20:28.578461: [debug] SELECT "id","action","timestamp" FROM "last_performed" WHERE "action"=?; [PersistInt64 1]
2020-05-04 04:20:28.578622: [debug] UPDATE "last_performed" SET "timestamp"=? WHERE "id"=? ; [PersistUTCTime 2020-05-04 04:20:28.577672 UTC,PersistInt64 1]
2020-05-04 04:20:28.578816: [debug] SELECT "id", "action", "timestamp" FROM "last_performed" WHERE "id"=? ; [PersistInt64 1]
2020-05-04 04:20:29.468533: [info] Current Stack version: 2.1.3, available download version: 2.3.1
2020-05-04 04:20:29.468669: [info] Newer version detected, downloading
2020-05-04 04:20:29.468718: [info] Querying for archive location for platform: linux-x86_64-static
2020-05-04 04:20:29.468777: [info] Downloading from: https://github.com/commercialhaskell/stack/releases/download/v2.3.1/stack-2.3.1-linux-x86_64-static.tar.gz
2020-05-04 04:20:53.308249: [info] Download complete, testing executable
Version 2.3.1, Git revision de2a7b694f07de7e6cf17f8c92338c16286b2878 (8103 commits) x86_64 hpack-0.33.0
2020-05-04 04:20:53.320343: [info] New stack executable available at /root/.local/bin/stack
2020-05-04 04:20:53.320455: [info] Also copying stack executable to /usr/local/bin/stack
2020-05-04 04:20:53.359993: [info] Stack executable copied successfully!
```
